File: 11.Deteccion_de_Bordes/test_2.m
clc; clear; close all
pkg load image
A=imread('images/linea3.jpg');
%Convertir la imagen en binaria
B=im2double(A);
B(B<0.5)=0; B(B>=0.5)=1;
imshow(B)

%Calcular la discretización de theta y rho
%1. Theta: Tomar valores en [0,180] / [0, pi]
h1=1;
thetas=deg2rad(0:h1:180);
%2. Rho: Toma valores en [-d,d], donde d=sqrt(m^2+n^2), y [m,n]=size(B)
[m,n]=size(B);
d=sqrt(m^2+n^2);
h2=1;
rhos=-d:h2:d;

%Crear matriz de acumulación
Acumulador=zeros(length(thetas),length(rhos));

%Llenar la matriz de acumulación

[x_b,y_b]=find(B);

for i=1:length(x_b) %Recorrer los puntos del borde
  for theta_ind=1:length(thetas)
    theta=thetas(theta_ind);
    rho=x_b(i)*cos(theta)+y_b(i)*sin(theta);
    [~,rho_ind]=min(abs(rhos-rho));
    Acumulador(theta_ind,rho_ind)+=1;
  end  
end

%Mostrar graficamente el comportamiento del acumulador
%figure
%surface(thetas,rhos,Acumulador','EdgeColor','none')
%xlabel('rho')
%ylabel('theta')

%Encontrar la posición donde el Acumulador alcanza el máxima
[xp,yp]=find(Acumulador==max(max(Acumulador)));
%Observación: Si el maximo se repite mas de una vez, entonces
%xp y yp son vectores que tienen las posiciones de esos valores

for k=1:length(xp)
    
     %Graficar
     thetaMax = thetas(xp(k));
     rhoMax = rhos(yp(k));
     
     if abs(sin(thetaMax)) < 10^-5 %Linea vertical
          xv = rhoMax/cos(thetaMax);
          line([n 1], [xv xv], "LineWidth",2)
     else
          pendiente = -cos(thetaMax)/sin(thetaMax);
          interseccion = rhoMax/sin(thetaMax);
          
          %Para graficar una linea recta, se utiliza el comando line
          %Necesitamos un dos puntos
          %Punto1 (1,y1)
          y1 = pendiente*1 + interseccion;
          %Punto2 (xn,n)
          xn = (n-interseccion)/pendiente;
     end

end



%Graficar la linea rectangle
line([y1 1], [1 xn], "LineWidth", 2)

File: 11.Deteccion_de_Bordes/test_2.py
import numpy as np
import matplotlib.pyplot as plt
from PIL import Image

A = Image.open("images/linea3.jpg")
A = np.asarray(A).astype("float64")/255
A = A[:,:,0]
A[A>=0.5]=1
A[A<0.5]=0
plt.imshow(A, cmap=plt.cm.gray)

h1 = 1
thetas = np.deg2rad(np.arange(0,180+h1,h1))

m,n = np.shape(A)
d = np.sqrt(m**2 + n**2)
h2 = 1
rhos = np.arange(-d,d,h2)

Acumulador = np.zeros((len(thetas),len(rhos)))
x_borde,y_borde = np.nonzero(A)

for i in range(len(x_borde)):
    for theta_indice in range(len(thetas)):
        theta = thetas[theta_indice]
        rho = x_borde[i]*np.cos(theta) + y_borde[i]*np.sin(theta)
        rho_indice = np.argmin(np.abs(rhos-rho))
        Acumulador[theta_indice,rho_indice] += 1

        #HAY UN PROBLEMA AQUI
        #XP RETORNA UN UNICO NUMERO Y DEBERIA RETORNAR UN ARRAY
        xp,yp = np.where(Acumulador == np.max(np.max(Acumulador)))

print(xp, yp)
y1 = 0
xn = 0

for k in range(len(xp)):
    thetaMax = thetas[k]
    rhoMax = rhos[k]

    if (np.abs(np.sin(thetaMax)) < 10**-5):
        xv = rhoMax / np.cos(thetaMax)
        plt.plot([n, 1], [xv, xv])
    else:
        pendiente = -np.cos(thetaMax) / np.sin(thetaMax)
        interseccion = rhoMax / np.sin(thetaMax)

        y1 = pendiente*1 + interseccion
        xn = (n-interseccion)/pendiente


plt.plot([y1, n], [1, xn])

plt.xlim([0,n])
plt.ylim([m,0])
plt.show() 